### ISSUE TITLE
Multi Core / Parallel Diffing

### ISSUE THREAD

**Issue by AlexLourenco** [joxeankoret/diaphora#138]
Hey @joxeankoret, I recently found out about this project and it's been a great help ever since I started to use it. I do a lot of work with old legacy versions of specific binaries to which I don't have access to the source. With that being said some of these binaries are quite large, taking around 3 to 4 hours to diff (with slow heuristics). I took a quick look at the source code and after indexing all heuristics are done as SQL queries, correct? What would the best way to add some concurrency into the equation? Would forking the process and splitting the different heuristics into separate process help? I'm not entirely sure how restrictive SQLite is on that regard, I presume the entire dataset is not loaded onto memory so I might have to either split parts of the database into separate files or straight up duplicate the entire database file in order to get truly concurrent reads? In any case I thought I would check with you before doing any work on this regard just to ensure that I am on the right path. 

As a quick disclaimer I'm not making this issue a feature request, I am more than happy to work on this myself and do a pull request once I'm done :)

(TLDR: I have a 40 core Xeon at my disposal. If I could shave off some of the time by putting it to work it would be great!)

**Comment by joxeankoret** [comment]
Hi @AlexLourenco!

Pretty much all heuristics are SQL queries, yeah. As for which would be the best way to add concurrency, I think first I need to change a bit how Diaphora launches the heuristics internally and, also, how are results selected. For example, let's say that we run at the same time 2 heuristics:

 * Heuristic A finds a match between funcA_1 and funcB_1.
 * At the same time, heuristic B finds the same match.

Right now, heuristics are executed in order of reliability: the most reliable ones first, then the less unreliable ones. Also, for each new heuristic being executed, all previous matches are discarded. If we add multiprocessing for launching heuristics, all rows would be always considered at the very least for the first simulatenously running heuristics. Perhaps these problems could be fixed like I do in Pigaios by saving all the multi-matches and then selecting the best ones + creating an in-memory SQLite table used as filter where I insert all results found by heuristics.

In summary: I like adding multi-processing but it requires some big changes from my side. Please give me a few days as I'm travelling and I'll take a look about how to implement parallelism in Diaphora.

**Comment by joxeankoret** [comment]
The first BETA version of parallel diffing is already working. Do you want to get a copy of the 2nd BETA version to test  @AlexLourenco?

**Comment by joxeankoret** [comment]
Done!
